### ISSUE TITLE
Error when storing actors reminders on MongoDB. error in updating document: an inserted document is too large

### ISSUE THREAD

**Issue by adejongh93** [dapr/dapr#6497]
## In what area(s)?

> /area runtime

## What version of Dapr?

1.11.0

## Expected Behavior

All actors reminders should be stored on MongoDB successfully

## Actual Behavior

Some actors reminders can not be stored on MongoDB due to very long value of field `data.$binary.base64`

## Steps to Reproduce the Problem

Create a simple dapr/dot-net application with MongoDB as state store, with one single actor and register many (40 should be enough) actor reminders.
Every time a new reminder is registered it will be added to the array of reminders on MongoDB. The value of `data.$binary.base64` will grow in length with every new reminder and will eventually exceed the size limit for a document in MongoDB ([16MB](https://www.mongodb.com/docs/manual/core/document/#:~:text=Document%20Size%20Limit,MongoDB%20provides%20the%20GridFS%20API.)). Every new reminder is appended at the end of the array, but the field `data.$binary.base64` will change everytime. At any moment, that field, in the reminder of position `i` of the array, when decoded with base64, will recursively contain the value of the same field in the reminder of position `i+1` of the array. After 40 reminders are registered, that value is large enough not to be allowed to be stored in MongoDB, and the exception `error in updating document: an inserted document is too large` is thrown.

[This repo](https://github.com/MargaretKrutikova/DaprReminderRepro) should help to reproduce this issue. Reminders registration can be triggered using swagger UI.

For example, in the code below we show the result after 3 reminders were registered.
The `base64` field of the first reminder, when decoded it returns this: `{"Subtype":0,"Data":"eyJTdWJ0eXBlIjowLCJEYXRhIjoiSWlJPSJ9"}`. The value of the second field of the result `"Data"` is the exact value of the  `base64` field of the second reminder. The same happens for the second and the third reminders

```
[{
  "_id": "actors||ActorTest",
  "_etag": "8a691475-3874-4b4b-975d-17bf6bb69dc4",
  "value": [
    {
      "actorid": "1",
      "actortype": "ActorTest",
      "name": "RandomReminder",
      "data": {
        "$binary": {
          "base64": "eyJTdWJ0eXBlIjowLCJEYXRhIjoiZXlKVGRXSjBlWEJsSWpvd0xDSkVZWFJoSWpvaVNXbEpQU0o5In0=",
          "subType": "00"
        }
      },
      "period": {},
      "registeredtime": {
        "$date": "2023-06-11T20:27:32.000Z"
      },
      "duetime": "1h0m0s0ms",
      "expirationtime": {
        "$date": {
          "$numberLong": "-62135596800000"
        }
      },
      "callback": ""
    },
    {
      "actorid": "2",
      "actortype": "ActorTest",
      "name": "RandomReminder",
      "data": {
        "$binary": {
          "base64": "eyJTdWJ0eXBlIjowLCJEYXRhIjoiSWlJPSJ9",
          "subType": "00"
        }
      },
      "period": {},
      "registeredtime": {
        "$date": "2023-06-11T20:27:37.000Z"
      },
      "duetime": "1h0m0s0ms",
      "expirationtime": {
        "$date": {
          "$numberLong": "-62135596800000"
        }
      },
      "callback": ""
    },
    {
      "actorid": "3",
      "actortype": "ActorTest",
      "name": "RandomReminder",
      "data": {
        "$binary": {
          "base64": "IiI=",
          "subType": "00"
        }
      },
      "period": {},
      "registeredtime": {
        "$date": "2023-06-11T20:28:45.000Z"
      },
      "duetime": "1h0m0s0ms",
      "expirationtime": {
        "$date": {
          "$numberLong": "-62135596800000"
        }
      },
      "callback": ""
    }
  ],
  "_ttl": null
}]
```

**Comment by MargaretKrutikova** [comment]
This bug might even be present in v 1.10.7, but we can't properly test because registering reminders fails with deserialization error on `expirationTime` field and it seems like only the first reminder is registered. v1.10.0 seems to at least not contain the serialization issue with the binary data in reminders. I suspect the bug might have been introduced in v1.10.6 in #6039 because some of the changes deal with marshalling/unmarshalling of the reminders metadata.

**Comment by artursouza** [comment]
Hi. Please, configure reminder partitioning to workaround database item size limits: https://docs.dapr.io/developing-applications/building-blocks/actors/howto-actors-partitioning/

**Comment by MargaretKrutikova** [comment]
No, that is not going to help. The bug is that the code reads the binary data from the data field without decoding it and encodes it again, making it grow every time a new reminder is added. It’s useless data and it’s not working as intended. Version 1.10.2 doesn’t have this problem.

**Comment by MargaretKrutikova** [comment]
My wildest guess would be that binary data is not decoded correctly with `json.Unmarshal` [here](https://github.com/dapr/dapr/blob/master/pkg/actors/actors.go#L1604) for the reminder `data` field which is stored as a bson binary field and seems to be base64 encoded. As a result every time a new reminder is added, the current list of reminders is fetched [here](https://github.com/dapr/dapr/blob/master/pkg/actors/actors.go#L1817) where the `data` field is not correctly decoded, and later the list is stored again adding a new level of base64 encoding to all `data` fields. Thus every new reminder adds a new "level" of base64 decoding to all the existing reminders' `data` field, and this is exactly of we are observing in that small repro use case.
Now, I am not sure who does the base64 serialization in this case, the actual storing of the reminder I believe is done in `dapr-contrib` in the underlying mongodb state store. Does any of this make sense?

**Comment by yaron2** [comment]
> My wildest guess would be that binary data is not decoded correctly with `json.Unmarshal` [here](https://github.com/dapr/dapr/blob/master/pkg/actors/actors.go#L1604) for the reminder `data` field which is stored as a bson binary field and seems to be base64 encoded. As a result every time a new reminder is added, the current list of reminders is fetched [here](https://github.com/dapr/dapr/blob/master/pkg/actors/actors.go#L1817) where the `data` field is not correctly decoded, and later the list is stored again adding a new level of base64 encoding to all `data` fields. Thus every new reminder adds a new "level" of base64 decoding to all the existing reminders' `data` field, and this is exactly of we are observing in that small repro use case.
> 
> The base64 serialization in this case seems to be done in [MarshalJSON](https://github.com/dapr/dapr/blob/master/pkg/actors/requests_responses.go#L84), however I don't see it applied where the actor reminders are fetched, there it is done via simple `json.Unmarshal`. Shouldn't some special unmarshalling be done? I see this [UnmarshalJSON](https://github.com/dapr/dapr/blob/master/pkg/actors/reminders/reminder_track.go#L47) for reminders, but I don't see it being used anywhere.

A JSON.Unmarshal method will trigger UnmarshalJSON, same thing for Marshal and MarshalJSON.

**Comment by artursouza** [comment]
Was this reproduced in 1.10 too? We will fix it for 1.11 and then 1.10 too if needed.

**Comment by MargaretKrutikova** [comment]
No, v.1.10.0 works well. It base64 encodes the data field properly and stores it as string. In one of the latest versions the data field became binary, not sure exactly which one, but can check it if it necessary.

**Comment by artursouza** [comment]
We had patches after 1.10.0. Does it work with 1.10.7 too?

**Comment by MargaretKrutikova** [comment]
It doesn't work for 1.10.7, no. We downgraded to 1.10.2 where everything with the reminders and mongodb works, we did have to make some serialization tricks (converting ints and enums to strings) to make it work, but eventually it did.

**Comment by berndverst** [comment]
The issue was introduced in Dapr 1.10.5 and has to do with the fact that Mongo insists on using BSON instead of JSON for its data serialization. We addressed this now by adding a custom BSON serialization function for ActorReminder data.

This will be included in 1.11.1 and 1.10.8 when they are being released.

**Comment by ItalyPaleAle** [comment]
Fixed by #6525 in release-1.11. Fix is being backported in other branches and will be included in 1.11.1 and 1.10.8

**Comment by mukundansundar** [comment]
in 1.11 #6525
